### PR TITLE
[100% tested] Ports multiple changes by goonstation to their process scheduler, These should cause a minor performance increase.

### PR DESCRIPTION
**Pull request by F-OS**
Hopefully this works correctly.
:cl:
TWEAK: Ports more recent code from the goon process scheduler.
TWEAK: Explosion processing now has a lag check, let coders know if this has any issues.
:cl:

### CODE REVIEW

File: code/game/objects/explosion_recursive.dm [hunk 1]
```diff
@@ -36,6 +36,7 @@ proc/explosion_rec(turf/epicenter, power)
 	for(var/direction in cardinal)
 		var/turf/T = get_step(epicenter, direction)
 		T.explosion_spread(power - epicenter.explosion_resistance, direction)
+		LAGCHECK(75)
```
[the remark this answers is not in the document]
Comment (F-OS) [reply code/game/objects/explosion_recursive.dm]: So port CHECK_LAG, LAG_CHECK, and STOP_LAG from TG?

[the remark this answers is not in the document]
Comment (MrStonedOne) [reply code/game/objects/explosion_recursive.dm]: I was mainly referring to the value, `75`

These values are for the same thing, you can port any of them to use in these lagchecks.
```dm
#define TICK_LIMIT_RUNNING 80
#define TICK_LIMIT_TO_RUN 70
#define TICK_LIMIT_MC 70
#define TICK_LIMIT_MC_INIT_DEFAULT 98
```

but tick_check would be better.

[the remark this answers is not in the document]
Comment (F-OS) [reply code/game/objects/explosion_recursive.dm]: Right, i'll port the tick limits now and work on the tick_check tomorrow, thanks!

**Comment by MrStonedOne** [comment]
When you decide to port stoplag, here is a mostly de-stonedmc-ified version

```dm
//Key thing that stops lag. Cornerstone of performance in ss13, Just sitting here, in unsorted.dm.

//Increases delay as the server gets more overloaded,
//as sleeps aren't cheap and sleeping only to wake up and sleep again is wasteful
#define DELTA_CALC max(max(world.tick_usage, world.cpu) / 100, 1)

//returns the number of ticks slept
/proc/stoplag(initial_delay)
	//replace with something to determine if initializations have happened
	if (!Master || !(Master.current_runlevel & RUNLEVELS_DEFAULT))
		sleep(world.tick_lag)
		return 1
	if (!initial_delay)
		initial_delay = world.tick_lag
	. = 0
	var/i = DS2TICKS(initial_delay) //get from _defines/math.dm
	do
		. += CEILING(i*DELTA_CALC, 1)
		sleep(i*world.tick_lag*DELTA_CALC)
		i *= 2
	while (world.tick_usage > TICK_LIMIT_TO_RUN)

#undef DELTA_CALC
```

**Comment by F-OS** [comment]
Thanks mate!

File: code/controllers/ProcessScheduler/process.dm [hunk 2]
```diff
@@ -303,11 +324,12 @@ datum/controller/process/proc/copyStateFrom(var/datum/controller/process/target)
 datum/controller/process/proc/onKill()
 
 datum/controller/process/proc/onStart()
+	LAGCHECK(world.tick_usage > 100 - tick_allowance)
```
Comment (MrStonedOne) [inline code/controllers/ProcessScheduler/process.dm]: This doesn't look right.

Comment (F-OS) [reply code/controllers/ProcessScheduler/process.dm]: This is a port from goon so i'll have to ask them

Comment (F-OS) [reply code/controllers/ProcessScheduler/process.dm]: @MrStonedOne 
capture

**Comment by Rohesie** [comment]
Didn't break the server while testmerged. There's that.

**Comment by Rohesie** [comment]
Resolve conflicts, please.

**Pull request merged**